src/main/java/de/uol/pgdoener/civicsage/search/SearchService.java

```diff
@@ -3,39 +3,65 @@
 import de.uol.pgdoener.civicsage.business.dto.SearchQueryDto;
 import de.uol.pgdoener.civicsage.business.dto.SearchResultDto;
 import de.uol.pgdoener.civicsage.mapper.SearchResultMapper;
+import de.uol.pgdoener.civicsage.search.exception.NotEnoughResultsAvailableException;
 import lombok.RequiredArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 import org.springframework.ai.document.Document;
+import org.springframework.ai.vectorstore.SearchRequest;
 import org.springframework.ai.vectorstore.VectorStore;
 import org.springframework.stereotype.Service;
 
 import java.util.List;
+import java.util.Optional;
 
 @Slf4j
 @Service
 @RequiredArgsConstructor
 public class SearchService {
 
+    private static final int DEFAULT_PAGE_NUMBER = 0;
+    private static final int DEFAULT_PAGE_SIZE = 20;
+
     private final VectorStore vectorStore;
     private final SearchResultMapper searchResultMapper;
 
-    public List<SearchResultDto> search(SearchQueryDto query) {
+    public List<SearchResultDto> search(SearchQueryDto query, Optional<Integer> pageNumber, Optional<Integer> pageSize) {
         log.info("Searching for documents with query {}", query);
+        int pNumber = pageNumber.orElse(DEFAULT_PAGE_NUMBER);
+        int pSize = pageSize.orElse(DEFAULT_PAGE_SIZE);
+        int resultsToFetch = (pNumber + 1) * pSize;
+        log.debug("topK = {}", resultsToFetch);
 
         // use SearchRequest instead of String to allow for more complex queries
-        List<Document> documents = vectorStore.similaritySearch(query.getQuery());
+        SearchRequest searchRequest = buildSearchRequest(query, resultsToFetch);
+        List<Document> documents = vectorStore.similaritySearch(searchRequest);
 
         if (documents == null || documents.isEmpty()) {
             log.info("No documents found for query {}", query);
             return List.of();
         }
 
         log.info("Found {} documents for query {}", documents.size(), query);
-        for (Document document : documents) {
-            log.debug("Document: {}", document);
-        }
+        documents = applyPagination(documents, pNumber, pSize);
 
         return searchResultMapper.toDto(documents);
     }
 
+    private SearchRequest buildSearchRequest(SearchQueryDto query, int resultsToFetch) {
+        return SearchRequest.builder()
+                .query(query.getQuery())
+                .topK(resultsToFetch)
+                .build();
+    }
+
+    private List<Document> applyPagination(List<Document> documents, int pageNumber, int pageSize) {
+        int startIndex = ((pageNumber + 1) * pageSize) - pageSize;
+        log.debug("startIndex = {}", startIndex);
+
+        if (startIndex > documents.size())
```

Comment:
```suggestion
        if (startIndex >= documents.size())
```

Without this I can still get an empty set as result.

![image](https://github.com/user-attachments/assets/81cc0c5a-9745-47b2-9367-6e92450d3a65)

Steps to reproduce:
- index something
- get number of indexed documents
- request page number 1 and pageSize equal to number of documents in vectorStore
- Actually, you just have to make the product of page number and page size equal to the number of documents in the vector store and you will always get an empty set.